plugins/json/json.cxx

```diff
@@ -0,0 +1,391 @@
+			{ // scalar member of the record
+				nlohmann::json result;
+				write_scalar_to_json(result, Ref_r{reference[i]}, logger);
+				json_data.push_back(result);
+			} else if (const auto&& sub_type = std::dynamic_pointer_cast<const Array_datatype>(tuple_type->elements()[i].type()))
+			{ // scalar member of the record
+				nlohmann::json result;
+				write_array_to_json(result, Ref_r{reference[i]}, logger);
+				json_data.push_back(result);
+			} else if (const auto&& sub_type = std::dynamic_pointer_cast<const Pointer_datatype>(tuple_type->elements()[i].type()))
+			{ // scalar member of the record
+				nlohmann::json result;
+				write_pointer_to_json(result, Ref_r{reference[i]}, logger);
+				json_data.push_back(result);
+			} else if (const auto&& sub_type = std::dynamic_pointer_cast<const Record_datatype>(tuple_type->elements()[i].type()))
+			{ // scalar member of the record
+				nlohmann::json result;
+				write_record_to_json(result, Ref_r{reference[i]}, logger);
+				json_data.push_back(result);
+			} else {
+				throw Type_error{"Unknown tuple subtype passed to json, currently supprting scalar and array subtypes."};
+			}
+		}
+	}
+
+	/** Write to json a pointer data
+	 *
+	 * \param json_data A json data to which we write a pointer data
+	 * \param reference A reference to a tuple datatype
+	 * \param logger PDI's logger instance
+	 */
+	void write_pointer_to_json(nlohmann::json& json_data, Ref_r reference, Logger& logger)
+	{
+		Ref dereferenced_ref = reference.dereference();
+		if (!dereferenced_ref) throw Value_error{"Can't dereference with read permissions"};
+		choose_type_and_dump_to_json(json_data, dereferenced_ref, logger);
+	}
+
+	/** Call different json dump function accordint to the datatype
+	 *
+	 * \param json_data A json data to which we write various types of data
+	 * \param reference A reference to a datatype
+	 * \param logger PDI's logger instance
+	 */
+	void choose_type_and_dump_to_json(nlohmann::json& json_data, Ref_r reference, Logger& logger)
+	{
+		if (const auto&& scalar_type = std::dynamic_pointer_cast<const Scalar_datatype>(reference.type())) { // a scalar type
+			write_scalar_to_json(json_data, reference, logger);
+		} else if (const auto&& array_type = std::dynamic_pointer_cast<const Array_datatype>(reference.type())) { // an array type
+			write_array_to_json(json_data, reference, logger);
+		} else if (const auto&& record_type = std::dynamic_pointer_cast<const Record_datatype>(reference.type())) { // a record type
+			write_record_to_json(json_data, reference, logger);
+		} else if (const auto&& pointer_type = std::dynamic_pointer_cast<const Tuple_datatype>(reference.type())) {
+			write_tuple_to_json(json_data, reference, logger);
+		} else if (const auto&& pointer_type = std::dynamic_pointer_cast<const Pointer_datatype>(reference.type())) { // a pointer type
+			write_pointer_to_json(json_data, reference, logger);
+		} else {
+			throw Type_error{"Unknown datatype passed to json"};
+		}
```

Comment:
I believe we can avoid all the if/else branches with templates.
Something like this:
```c++
json_data = write_to_json<reference.type()>(reference);
```
And then have specializations per type. A little like this:
```c++
template<>
nlohmann::json write_to_json<const Scalar_datatype>(Reference&) {
// TODO
}
```

Reply:
This comment also applies to other if/else blocks.

Reply:
not feasible for the moment